nemo/collections/asr/inference/pipelines/base_pipeline.py

```diff
@@ -0,0 +1,363 @@
+    """
+
+    def __init__(self, texts: list[str] | None = None, segments: list[list[TextSegment]] | None = None):
+        """
+        Initialize the PipelineOutput.
+        Args:
+            texts: (list[str] | None) List of transcriptions.
+            segments: (list[list[TextSegment]] | None) List of corresponding text segments.
+        """
+        if texts is None and segments is None:
+            raise ValueError("At least one of the 'texts' or 'segments' should be provided.")
+        self.texts = texts
+        self.segments = segments
+
+
+class BasePipeline(PipelineInterface):
+    """
+    Base class for all pipelines.
+    """
+
+    def __init__(self):
+        """Initialize state pool to store the state for each stream"""
+        self._state_pool: dict[int, Any] = {}
+
+    def get_state(self, stream_id: int) -> Any:
+        """Retrieve state for a given stream ID."""
+        return self._state_pool.get(stream_id, None)
+
+    def get_states(self, stream_ids: Iterable[int]) -> list[Any]:
+        """Retrieve states for a list of stream IDs."""
+        return [self.get_state(stream_id) for stream_id in stream_ids]
+
+    def delete_state(self, stream_id: int) -> None:
+        """Delete the state from the state pool."""
+        if stream_id in self._state_pool:
+            del self._state_pool[stream_id]
+
+    def delete_states(self, stream_ids: Iterable[int]) -> None:
+        """Delete states for a list of stream IDs."""
+        for stream_id in stream_ids:
+            self.delete_state(stream_id)
+
+    def init_state(self, stream_id: int, options: ASRRequestOptions) -> Any:
+        """Initialize the state of the stream"""
+        if stream_id not in self._state_pool:
+            state = self.create_state(options)
+            self._state_pool[stream_id] = state
+        return self._state_pool[stream_id]
+
+    def reset_session(self) -> None:
+        """Reset the frame buffer and internal state pool"""
+        self._state_pool.clear()
+
+    def open_session(self) -> None:
+        """Start a new session by resetting the internal state pool"""
+        self.reset_session()
+
+    def close_session(self) -> None:
+        """Close the session by resetting the internal state pool"""
+        self.reset_session()
+
+    @abstractmethod
+    def transcribe_step_for_frames(self, frames: list[Frame]) -> None:
+        """Transcribe a step for frames"""
+        pass
+
+    @abstractmethod
+    def transcribe_step_for_feature_buffers(self, fbuffers: list[FeatureBuffer]) -> None:
+        """Transcribe a step for feature buffers"""
+        pass
+
+    @abstractmethod
+    def get_request_generator(self) -> ContinuousBatchedRequestStreamer:
+        """Return the request generator."""
+        pass
+
+    @abstractmethod
+    def get_sep(self) -> str:
+        """Return the separator for the text postprocessor."""
+        pass
+
+    def transcribe_step(self, requests: list[Request]) -> None:
```

Comment:
Do we have the option to let transcribe step returns the results of current step? When we post a request to the inference server, we usually expect to get the result for that request, returning `None` seems a bit strange...

Reply:
Fixed